File: 백준/Gold/6593. 상범 빌딩/상범 빌딩.java
import java.io.*;
import java.util.*;

public class Main {

    static int L, R, C;
    static char[][][] building;
    static boolean[][][] visited;
    static int[] dx = {0, 1, 0, -1, 0, 0};  // 동, 남, 서, 북, 상, 하
    static int[] dy = {1, 0, -1, 0, 0, 0};
    static int[] dz = {0, 0, 0, 0, 1, -1};  // 층 이동은 z축

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        while (true) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            L = Integer.parseInt(st.nextToken());
            R = Integer.parseInt(st.nextToken());
            C = Integer.parseInt(st.nextToken());

            if (L == 0 && R == 0 && C == 0) break;

            building = new char[L][R][C];
            visited = new boolean[L][R][C];

            int startZ = 0, startX = 0, startY = 0;

            for (int l = 0; l < L; l++) {
                for (int r = 0; r < R; r++) {
                    String line = br.readLine();
                    for (int c = 0; c < C; c++) {
                        building[l][r][c] = line.charAt(c);
                        if (building[l][r][c] == 'S') {
                            startZ = l;
                            startX = r;
                            startY = c;
                        }
                    }
                }
                br.readLine(); // 빈 줄 처리
            }

            int result = bfs(startZ, startX, startY);
            if (result == -1) {
                System.out.println("Trapped!");
            } else {
                System.out.println("Escaped in " + result + " minute(s).");
            }
        }
    }

    static int bfs(int z, int x, int y) {
        Queue<int[]> queue = new ArrayDeque<>();
        queue.add(new int[]{z, x, y, 0});
        visited[z][x][y] = true;

        while (!queue.isEmpty()) {
            int[] cur = queue.poll();
            int cz = cur[0], cx = cur[1], cy = cur[2], time = cur[3];

            if (building[cz][cx][cy] == 'E') return time;

            for (int i = 0; i < 6; i++) {
                int nz = cz + dz[i];
                int nx = cx + dx[i];
                int ny = cy + dy[i];

                if (0 <= nz && nz < L && 0 <= nx && nx < R && 0 <= ny && ny < C) {
                    if (!visited[nz][nx][ny] && building[nz][nx][ny] != '#') {
                        visited[nz][nx][ny] = true;
                        queue.add(new int[]{nz, nx, ny, time + 1});
                    }
                }
            }
        }

        return -1;
    }
}

File: 백준/Gold/6593. 상범 빌딩/상범 빌딩.py
import sys
from collections import deque

di = [0, 1, 0, -1]
dj = [1, 0, -1, 0]
df = [-1, 1]

def bfs(f, i, j):
    q = deque()
    q.append([f, i, j, 0]) # 층수, x, y 좌표
    visited[f][i][j] = True

    while q:
        floor, x, y, cnt = q.popleft()

        if building[floor][x][y] == "E":
            return cnt

        for k in range(4):
            nx, ny = x + di[k], y + dj[k]
            if 0 <= nx < R and 0 <= ny < C and not visited[floor][nx][ny] and building[floor][nx][ny] != "#":
                q.append([floor, nx, ny, cnt + 1])
                visited[floor][nx][ny] = True

        for k in range(2):
            next_floor = floor + df[k]
            if 0 <= next_floor < L and not visited[next_floor][x][y] and building[next_floor][x][y] != "#":
                q.append([next_floor, x, y, cnt+1])
                visited[next_floor][x][y] = True
    return -1
while True:
    L, R, C = map(int, input().split())
    if L == 0 and R == 0 and C == 0:
        break

    building = []
    start = None
    for l in range(L):
        floor = []
        for r in range(R):
            row = list(input().strip())
            for c in range(C):
                if row[c] == "S":
                    start = (l, r, c)
            floor.append(row)
        building.append(floor)
        input()

    visited = [[[False] * C for _ in range(R)] for _ in range(L)]

    ans = bfs(*start)

    if ans == -1:
        print("Trapped!")
    else:
        print(f"Escaped in {ans} minute(s).")


